### ISSUE TITLE
[Bug]: Invalid code in Stream._increment_stream_state

### ISSUE THREAD

**Issue by laurentS** [meltano/sdk#1161]
### Singer SDK Version

0.13.1

### Python Version

3.10

### Bug scope

Taps (catalog, state, stream maps, etc.)

### Operating System

linux

### Description

I think

https://github.com/meltano/sdk/blob/main/singer_sdk/streams/core.py#L722:
```python
if not treat_as_sorted and self.state_partitioning_keys is not None:
    # Streams with custom state partitioning are not resumable.
    treat_as_sorted = False
```

is incorrect and should be written as
```python
if treat_as_sorted and self.state_partitioning_keys is not None:
```
With the current code, this is basically a no-op, as the test can only pass if `treat_as_sorted` is already `False`.

### Code

```python
I noticed this while reading the code to figure out something else. I have no buggy code to show :)

I guess there is probably no test covering this particular case (stream with a `state_partitioning_keys` and `is_sorted = True`), otherwise, this would have surfaced.
```

**Comment by tayloramurphy** [comment]
@edgarrmondragon is this still a bug?

**Comment by edgarrmondragon** [comment]
> @edgarrmondragon is this still a bug?

@tayloramurphy Well, that piece of code's still out there and it is indeed a no-op, but I'm not sure what buggy behavior this produces without taking a deeper look.

**Comment by aaronsteers** [comment]
Hey, ya'll. 👋 

I saw the notification pop up on this and did a bit of digging. First, I do agree this appears to be a bug. The intended behavior is to treat streams as unsorted if they are overriding the stream state - specifically because _within_ the scope of `n` parents, if `n > 1`, then the results contained within do not represent a purely sorted dataset.

(The most common reason to override stream partitioning keys is to reduce the size of the state when number of parent items is >1000.)

After digging and tracing through a view different 'blame' records, I eventually found that apparently this "no-op bug" was introduced at the very onset of creating the parent-child streams feature. The PR itself isn't on GitHub because we were still on GitLab in those days (May 2021), but the Git commit is here:

- https://github.com/meltano/sdk/commit/8e4996d2bf70d6e31cf274684d535582d24a0560

As far as I can tell, this has always been a no-op. The code could be removed, or it could be "fixed" but that fix would be best paired with some testing. It's also possible there's another place in the code that's producing the desired effect, but I can't say whether or not that's the case.

I don't personally have time to put into a fix, but I did just want to provide the extra context in case it is helpful. I don't have a very strong recommendation on which path to take. But either option (removing the no-op code or fixing the code with some added tests to ensure desired behavior) both seem to be good options. Since no one seems to have logged an issue in the two years since that original commit, I'd probably lean towards removing the offending code - since it seems clear to be a no-op and removing shouldn't have any averse side effects. 😄 🤷 

Hope this helps!
